# 多维数组的运算

In [1]:
import numpy as np
import math

In [2]:
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
], dtype=np.float64)
B = np.array([
    [3,2,1],
    [6,5,4],
    [9,8,7],
], dtype=np.float64)

## 普通运算

普通的运算符将会依次计算数组中的对应位：

In [3]:
A + B

array([[ 4.,  4.,  4.],
       [10., 10., 10.],
       [16., 16., 16.]])

In [4]:
A - B

array([[-2.,  0.,  2.],
       [-2.,  0.,  2.],
       [-2.,  0.,  2.]])

In [5]:
A * B

array([[ 3.,  4.,  3.],
       [24., 25., 24.],
       [63., 64., 63.]])

In [6]:
A / B

array([[0.33333333, 1.        , 3.        ],
       [0.66666667, 1.        , 1.5       ],
       [0.77777778, 1.        , 1.28571429]])

In [7]:
A ** B

array([[1.0000000e+00, 4.0000000e+00, 3.0000000e+00],
       [4.0960000e+03, 3.1250000e+03, 1.2960000e+03],
       [4.0353607e+07, 1.6777216e+07, 4.7829690e+06]])

In [8]:
A % B

array([[1., 0., 0.],
       [4., 0., 2.],
       [7., 0., 2.]])

## 布尔运算

布尔运算将会返回真值表，这在用于过滤数组中的某些数据时非常有用。

In [9]:
A < B

array([[ True, False, False],
       [ True, False, False],
       [ True, False, False]])

## 常用统计方法

这些统计方法会将任何维的数组视作一维数组对待，所以不需要提前 `reshape`。
实际上， numpy 底层也只有一维数组，只是根据 shape 属性的不同，进行不同的运算方法而已。

- 最大值： `max()`
- 最小值： `min()`
- 累和： `sum()`
- 累积： `prod()`

In [10]:
A.max()

9.0

In [11]:
A.min()

1.0

In [12]:
A.sum()

45.0

In [13]:
A.prod()

362880.0

关于累加和累积，还分别有个带 `cum` 前缀的版本，其将会保留遍历计算的每一步。

In [14]:
A.cumsum()

array([ 1.,  3.,  6., 10., 15., 21., 28., 36., 45.])

In [15]:
A.cumprod()

array([1.0000e+00, 2.0000e+00, 6.0000e+00, 2.4000e+01, 1.2000e+02,
       7.2000e+02, 5.0400e+03, 4.0320e+04, 3.6288e+05])

## 矩阵运算

二维数组可以用来模拟矩阵，在 `np.linalg` 中封装了许多线性代数相关的计算方法。

### 点乘

矩阵/向量 进行点乘使用 `np.dot` 方法, 在 Python 3.5 之后， `@` 被重载为点乘：

In [16]:
np.dot(A, B)

array([[ 42.,  36.,  30.],
       [ 96.,  81.,  66.],
       [150., 126., 102.]])

In [49]:
A @ B

array([[ 42.,  36.,  30.],
       [ 96.,  81.,  66.],
       [150., 126., 102.]])

$$
\begin{bmatrix}
1 & 2 & 3\\
4 & 5 & 6\\
7 & 8 & 9\\
\end{bmatrix} \cdot \begin{bmatrix}
3&2&1\\
6&5&4\\
9&8&7\\
\end{bmatrix} = \begin{bmatrix}
42&36&30\\
96&81&66\\
150&126&102\\
\end{bmatrix}
$$

In [27]:
np.dot(B, A)

array([[ 18.,  24.,  30.],
       [ 54.,  69.,  84.],
       [ 90., 114., 138.]])

In [22]:
# 列表自动构造为一维数组
np.dot(
[1, 2, 3], [1, 1, 1]
)

6

$$
\begin{bmatrix}
1\\ 2\\ 3
\end{bmatrix} \cdot \begin{bmatrix}
1\\ 1\\ 1
\end{bmatrix} = 1+2+3 = 6
$$

数组表示怎样的矩阵、向量，可以通过观察其 `.shape` 属性来决定。
但数组实际上并不是矩阵，在取切片之后，其形状会发生变化。其 shape 属性如果为 `(3,)` 则表示一维数组， `(1,3)` 表示 3 阶行向量， `(3,1)` 表示 3 阶列向量。

In [51]:
A.shape

(3, 3)

In [57]:
a = A[:, 1]
assert a.shape == (3,)
a

array([2., 5., 8.])

In [58]:
a = A[1, :]
assert a.shape == (3,)
a

array([4., 5., 6.])

由于一维数组的 shape 与列向量相近，所以在许多行为上和列向量一致，但某些情况下也会出现问题，例如，一个 (3, 2) 矩阵应当能与 (2, 1) 的列向量点乘，并得到 (3,1) 的列向量，但是对于数组就会报错：

In [ ]:
a = np.array([1, 2, 3])

### 叉乘

使用 `np.cross`

In [45]:
np.cross([1, 2, 3], [4, 5, 6])

array([-3,  6, -3])

$$
\begin{bmatrix}
1\\2\\3
\end{bmatrix} \times \begin{bmatrix}
4\\5\\6
\end{bmatrix} = \left|
\begin{array}? % 吞字占位
\vec{i}&\vec{j}&\vec{k}\\
1&2&3\\
4&5&6\\
\end{array} \right| =
\begin{bmatrix}
-3\\6\\-3
\end{bmatrix}
$$

### 转置

转置矩阵可以使用 `np.tranpose` 函数，也可以调用 ndarray 的 `T` 属性。

In [28]:
C = np.array([1, 2, 3])

In [29]:
C

array([1, 2, 3])

In [32]:
np.transpose(C)

array([1, 2, 3])

In [31]:
C.T

array([1, 2, 3])

### 逆矩阵

In [40]:
np.linalg.inv(A)

array([[ 3.15251974e+15, -6.30503948e+15,  3.15251974e+15],
       [-6.30503948e+15,  1.26100790e+16, -6.30503948e+15],
       [ 3.15251974e+15, -6.30503948e+15,  3.15251974e+15]])

### 行列式

In [41]:
np.linalg.det(A)

-9.51619735392994e-16

### 求秩

求方阵的秩，使用 `np.linalg.matrix_rank`，其背后使用的是 SVD 算法。

> SVD: 奇异值分解

In [42]:
np.linalg.matrix_rank(A)

2

### 求解线性方程组

在线性代数课上学的，求线性方程组：

1. 求系数矩阵的逆矩阵
2. 右值矩阵左乘此逆矩阵

这个步骤被包装为 `np.linalg.solve`

In [47]:
np.linalg.solve([[3, 1,],[1, 2]], [3, 2])

array([0.8, 0.6])

$$
\begin{bmatrix}
3&1\\
1&2\\
\end{bmatrix}\begin{bmatrix}
0.8\\
0.6
\end{bmatrix} = \begin{bmatrix}
3 \\ 2
\end{bmatrix}
$$